File: day8.js
import { readInput } from "./helper.js";

const input = readInput("./input/day8.txt");

/**
 * Run part 1 of day 8 problem
 */
function part1() {
  let acc = 0;
  let visitedIndexes = new Array(input.length);
  let index = 0;
  while (true) {
    if (visitedIndexes[index] === 1) {
      break;
    } else {
      //visit that index, and do the operation at that index in the input
      visitedIndexes[index] = 1;
      //get input value
      const op = input[index].split(" ");

      switch (op[0]) {
        case "nop":
          index++;
          break;
        case "acc":
          acc += parseInt(op[1]);
          index++;
          break;
        case "jmp":
          index += parseInt(op[1]);
      }
    }
  }
  console.log("Part 1:", acc);
}

/**
 * Run part 2 of day 8 problem
 */
function part2() {
  let acc = 0;
  let visitedIndexes = new Array(input.length);
  let index = 0;
  let jmp = [];
  let iteration = 1;
  while (true) {
    if (index >= input.length) {
      break;
    }
    if (visitedIndexes[index] === 1) {
      //reset and change next jump to nop
      acc = 0;
      visitedIndexes = new Array(input.length);
      index = 0;
      iteration++;
    } else {
      //visit that index, and do the operation at that index in the input
      visitedIndexes[index] = 1;
      //get input value
      const op = input[index].split(" ");

      switch (op[0]) {
        case "nop":
          index++;
          break;
        case "acc":
          acc += parseInt(op[1]);
          index++;
          break;
        case "jmp":
          if (jmp.length !== iteration) {
            if (!jmp.includes(index)) {
              jmp.push(index);
              index++;
            } else {
              index += parseInt(op[1]);
            }
          } else {
            index += parseInt(op[1]);
          }
      }
    }
  }
  console.log("Part 2:", acc);
}
part1();
part2();

File: 2020/day8.py
print('Python')
file = open('./input/day8.txt')
file = [x.strip() for x in file.readlines()]


def part1():
    acc = 0
    visitedIndexes = [-1] * len(file)
    index = 0
    while True:
        if visitedIndexes[index] == 1:
            break
        visitedIndexes[index] = 1
        op = file[index].split(" ")
        if 'nop' in op[0]:
            index += 1
        elif 'acc' in op[0]:
            acc += int(op[1])
            index += 1
        elif 'jmp' in op[0]:
            index += int(op[1])
    print('Part 1 : {}'.format(acc))


def part2():
    acc = 0
    visitedIndexes = [-1] * len(file)
    index = 0
    jmp = []
    iteration = 1
    while True:
        if index >= len(file):
            break
        if visitedIndexes[index] == 1:
            acc = 0
            visitedIndexes = [-1] * len(file)
            index = 0
            iteration += 1
        else:
            visitedIndexes[index] = 1
            op = file[index].split(" ")
            if 'nop' in op[0]:
                index += 1
            elif 'acc' in op[0]:
                acc += int(op[1])
                index += 1
            elif 'jmp' in op[0]:
                if len(jmp) != iteration:
                    if index not in jmp:
                        jmp.append(index)
                        index += 1
                    else:
                        index += int(op[1])
                else:
                    index += int(op[1])
    print('Part 2 : {}'.format(acc))


part1()
part2()